### PR TITLE
5 delete items

### LINKED ISSUE #5: Delete Items
Implement some way for the user to delete items in their list.

### PR DESCRIPTION
**Pull request by tycebrown**
Added a way to delete items. I'll be glad to hear your changes!

### CODE REVIEW

File: lib/model/utils.dart [hunk 1]
```diff
@@ -0,0 +1,12 @@
+/// Util:
+///   - a static class that provides
+///     various utilities
+class Utils {
+  static Iterable<U> mapIndexed<T, U>(
```
Comment (Pertempto) [inline lib/model/utils.dart]: It's not immediately obvious to me what this method does, I have to read it to figure it out. This means you need a better function name and/or method comment (directly above the method).

File: lib/model/utils.dart [hunk 2]
```diff
@@ -0,0 +1,12 @@
+/// Util:
```
Comment (Pertempto) [inline lib/model/utils.dart]: Instead of putting this in the `model` directory, I would probably put it in a `shared` or `common` directory.

File: lib/view/item_widget.dart [hunk 3]
```diff
@@ -6,7 +6,8 @@ import 'package:lists/model/item.dart';
 ///     in a list (see Item)
 class ItemWidget extends StatefulWidget {
   final Item item;
-  const ItemWidget(this.item, {super.key});
+  final void Function() deleteThisItem;
+  const ItemWidget(this.item, {required this.deleteThisItem, super.key});
```
Comment (Pertempto) [inline lib/view/item_widget.dart]:
```suggestion
  const ItemWidget(this.item, {required this.onDelete, super.key});
```

File: lib/view/item_widget.dart [hunk 4]
```diff
@@ -19,11 +20,40 @@ class _ItemWidgetState extends State<ItemWidget> {
...
+    showDialog(
+      context: context,
+      builder: (context) => AlertDialog(
+        title: const Text("Delete this item?", textAlign: TextAlign.center),
+        actionsAlignment: MainAxisAlignment.spaceBetween,
+        actions: [
+          IconButton(
+            onPressed: () {
+              Navigator.pop(context);
+            },
```
Comment (Pertempto) [inline lib/view/item_widget.dart]:
```suggestion
            onPressed: () => Navigator.pop(context),
```
Here's a shorter way to write that.  
Tip: the `Ctrl + .`  shortcut in VS code is very helpful for this kind of thing. Just put your cursor in a block of code and use that shortcut to get a helpful list of quick actions that the editor can do for you automatically! I use this shortcut a ton with Flutter development to save time.

File: lib/view/item_widget.dart [hunk 5]
```diff
@@ -19,11 +20,40 @@ class _ItemWidgetState extends State<ItemWidget> {
...
   }
 
+  void _showConfirmDeleteDialog() {
+    showDialog(
+      context: context,
+      builder: (context) => AlertDialog(
+        title: const Text("Delete this item?", textAlign: TextAlign.center),
+        actionsAlignment: MainAxisAlignment.spaceBetween,
+        actions: [
+          IconButton(
```
Comment (Pertempto) [inline lib/view/item_widget.dart]: Using a red icon to signify cancel is confusing. In a delete dialog, I would expect the red button to be the one that deletes the item. Instead of using `IconButton`, I would use `TextButton`. It is the normal button used with dialogs.

File: lib/view/item_widget.dart [hunk 6]
```diff
@@ -19,11 +20,40 @@ class _ItemWidgetState extends State<ItemWidget> {
   Widget build(BuildContext context) {
     final itemTextStyle = Theme.of(context).textTheme.titleLarge;
     return ListTile(
+      leading: IconButton(
```
Comment (Pertempto) [inline lib/view/item_widget.dart]: Don't put the delete button at the beginning of the item. It looks very strange. At the end of the item is better, but at the same time it doesn't look great to have so many delete icons on the screen.  
Do you have the Gmail or Google Messages app on Android? Experiment with them and see how they implement delete functionality. It is usually through some kind of press and hold action.  I think this would be complicated to implement, so we need something simpler.

Some ideas:
* Use the [`flutter_slidable`](https://pub.dev/packages/flutter_slidable) library to implement a slide gesture that displays the delete button.
* Use a modal bottom sheet ([docs here](https://api.flutter.dev/flutter/material/showModalBottomSheet.html)) to display the delete button. When the user presses and holds on an item, display the bottom sheet for the item. Currently the bottom sheet would only contain the delete button. In the future, you could add more buttons to the bottom sheet.

File: lib/view/item_widget.dart (hunk 5 above)
Comment (Pertempto) [reply lib/view/item_widget.dart]: Leave the cancel button normal colored, and make the delete button red.

**Review by Pertempto — CHANGES_REQUESTED** [review]
A few things I found...

File: lib/model/utils.dart (hunk 1 above)
Comment (Pertempto) [reply lib/model/utils.dart]: Really you don't need this method at all. It looks like it was copy and pasted from the `collection` library. Just add the library as a dependency and import it, instead of copying the code.

Comment (Pertempto) [reply lib/model/utils.dart:5]: https://pub.dev/packages/collection

**Review by Pertempto — CHANGES_REQUESTED** [review]

File: lib/model/utils.dart (hunk 1 above)
Comment (Pertempto) [reply lib/model/utils.dart]: https://pub.dev/packages/collection

**Comment by Pertempto** [comment]
I'm not going to use the "changes requested" label, because it seems that Github has good built-in support for displaying "Changes requested". I'm still getting use to using Github, it's not what I use on a day-to-day basis.

File: lib/view/item_widget.dart (hunk 6 above)
Comment (tycebrown) [reply lib/view/item_widget.dart]: I decided to go with your idea of a long press bringing up a modal bottom sheet.

File: lib/main.dart [hunk 7]
```diff
@@ -14,6 +14,7 @@ class MyApp extends StatelessWidget {
     return MaterialApp(
       title: 'Flutter Demo',
       theme: ThemeData(
+          iconTheme: IconThemeData.fallback().copyWith(size: 40),
```
Comment (Pertempto) [inline lib/main.dart]: Why did you add this?

File: lib/view/item_widget.dart [hunk 8]
```diff
@@ -19,11 +20,28 @@ class _ItemWidgetState extends State<ItemWidget> {
...
+      builder: (context) => Row(
+        children: [
+          TextButton(
+              onPressed: () {
+                widget.onDelete();
+                Navigator.pop(context);
+              },
+              child: const Icon(Icons.delete)),
+        ],
+      ),
```
Comment (Pertempto) [inline lib/view/item_widget.dart]:
```suggestion
    builder: (context) => Padding(
        padding: const EdgeInsets.all(8.0),
        child: Row(
          mainAxisAlignment: MainAxisAlignment.spaceBetween,
          children: [
            TextButton.icon(
                onPressed: () {
                  widget.onDelete();
                  Navigator.pop(context);
                },
                icon: const Icon(Icons.delete),
                label: const Text('Delete'),
                style: const ButtonStyle(
                    foregroundColor:
                        MaterialStatePropertyAll<Color>(Colors.red)))
          ],
        ),
      ),
```

Comment (Pertempto) [reply lib/view/item_widget.dart]: Add some padding around your content, so it doesn't look so squished. Use an icon with text, colored red to show that it is a "danger" action.

File: lib/main.dart (hunk 7 above)
Comment (Pertempto) [reply lib/main.dart]: It looks strange to have such large icons. The default size is fine.

**Review by Pertempto — CHANGES_REQUESTED** [review]

File: lib/view/item_widget.dart (hunk 8 above)
Comment (tycebrown) [reply lib/view/item_widget.dart]: That makes it look a lot better.

I am concerned, though, that if we add more buttons with labels, the modal sheet's row might overflow. Do you want me to look into adding some conditional visibility to the label?

Comment (Pertempto) [reply lib/view/item_widget.dart]: I wouldn't worry about it for now. Rows can wrap, they don't have to overflow. If we have to add more buttons, we can look at ways to show them all well then.

**Review by Pertempto — APPROVED** [review]
Great work!
